Add AbsorptionCoefficient and Emissivity to MeanOpacity variants.

RyanWollaeger opened the pull request:
+ Use gmode input to toggle Rosseland (default) to Planck.
+ Add PlanckDistribution function as member to MeanOpacity.
+ Add the new functions to generic and non-CGS MeanOpacity variants.

Note: the motivation is to use one of the frequency-integrated opacity values as a grey opacity in transport, which requires using it in both absorption and emissivity, for any choice.

RyanWollaeger committed:
Add AbsorptionCoefficient and Emissivity to MeanOpacity variants.

+ Use gmode input to toggle Rosseland (default) to Planck.
+ Add PlanckDistribution function as member to MeanOpacity.
+ Add the new functions to generic and non-CGS MeanOpacity variants.

Note: the motivation is to use one of the frequency-integrated
opacity values as a grey opacity in transport, which requires using
it in both absorption and emissivity, for any choice.

Yurlungur reviewed (approved):
This may eventually be something we want for neutrinos too... in which case, it might be good to add it there too. The API should be identical. That said, if you don't want to attack that right now, that's fine.

RyanWollaeger commented:
@brryan @Yurlungur @pdmullen - this is a preliminary attempt for something that might be used in grey transport. I added the Planck function as a member to the MeanOpacity class, which seems a bit overkill for just emissivity.

RyanWollaeger commented on singularity-opac/photons/mean_opacity_photons.hpp:
```diff
@@ -143,6 +144,25 @@ class MeanOpacity {
+  // standard grey opacity functions
+  PORTABLE_INLINE_FUNCTION
+  Real AbsorptionCoefficient(const Real rho, const Real temp,
+                            const int gmode = Rosseland) const {
```
Would it make sense to move this `gmode` index to the constructor? We may want to avoid different function signatures between non-grey mean opacities and mean opacities, so that they can be used interchangeably.

RyanWollaeger replied:
We decided this should be in the AbsorptionCoefficient and Emissivity arguments.

RyanWollaeger replied:
(Accidentally added this comment in a pending review)

brryan commented on singularity-opac/photons/mean_photon_variant.hpp:
```diff
@@ -92,6 +92,27 @@ class MeanVariant {
 
+  PORTABLE_INLINE_FUNCTION
+  Real AbsorptionCoefficient(const Real rho, const Real temp,
+                            const int gmode = 0) const {
```
```suggestion
                            const int gmode = Rosseland) const {
```
Is this change OK to make? There are a few other places with integers as default args to these functions here and below.

RyanWollaeger replied:
For some reason my comment got duplicated, I deleted the duplication, but that deleted it here - I made the change, but asked about `non_cgs_photons.hpp`, which doesn't include this as a header, but has these `gmode` values also defaulted to integers. Should the enum be moved to a separate header to be included here and in `non_cgs_photons.hpp`?

RyanWollaeger replied:
@brryan - any thoughts on this header question? Thanks!

brryan replied:
@RyanWollaeger Sorry I'm not sure how I missed this question. Yeah I think you're right, we need a new header file, `photon_types.hpp` or something (you may have a better name in mind) to be included in both these existing files.

Yurlungur replied:
Should this be mimicked in neutrinos too? It might be useful to have rosseland vs planck mean in neutrinos too

RyanWollaeger replied:
Just made this update for photons - will map this over to neutrinos shortly, along with the other AbsorptionCoefficient/Emissivity changes.

brryan replied:
We can try out just the photons for now and see if anything goes wrong bringing it into artemis before typing out the neutrinos

brryan reviewed (approved):
LGTM! I think the planck/rosseland choice is best made when calling the function, either as you do here with function args, or with different fucntions or a template parameter. What you have here is fine IMO, I just think it's better to not do this in the constructor because we dont want to make the same class twice to call the same data. I'm unconcerned about the different function signatures between mean and non-mean opacity variants, because they have different meanings.

RyanWollaeger committed:
Use Rosseland/Planck enum in photon MeanVariant.

+ Move enum from MeanOpacity class to mean_photon_variant.hpp.
+ Use Rosseland instead of 0 in variant template function defaults.

Yurlungur commented:
@RyanWollaeger what's the status of this MR?

brryan commented on singularity-opac/photons/non_cgs_photons.hpp:
```diff
@@ -246,7 +246,8 @@ class MeanNonCGSUnits {
         mass_unit_(mass_unit), length_unit_(length_unit), temp_unit_(temp_unit),
-        rho_unit_(mass_unit_ / (length_unit_ * length_unit_ * length_unit_)) {}
+        rho_unit_(mass_unit_ / (length_unit_ * length_unit_ * length_unit_)),
+        inv_emiss_unit_(length_unit_ * time_unit_ * time_unit_ / mass_unit_) {}
```
Is this not 
```suggestion
        inv_emiss_unit_(length_unit_ * time_unit_ * time_unit_ * time_unit_ / mass_unit_) {}
```
? since
`<J> = <dE / d^3x dt> = g cm^2 / (s^2 cm^3 s) = g / (cm s^3)`

brryan replied:
Oh I see you add the extra `time_unit_` in separately later... but any reason to not just include that here?

RyanWollaeger replied:
This may've been to conform to the pattern for the non-mean variant - will double-check.

RyanWollaeger replied:
Just made this change - thanks!

brryan replied:
🙌

RyanWollaeger committed:
Move mean opacity enum to its own header: mean_photon_types.hpp.

RyanWollaeger committed:
Roll time unit into mean non-cgs inverse emissivity unit.

brryan reviewed (approved):
LGTM! Great work @RyanWollaeger!
